prompt/select.ts

```diff
@@ -99,6 +146,167 @@ export class Select<S extends SelectSettings = SelectSettings>
     return this.options[this.listIndex]?.value ?? this.settings.default;
   }
 
+  protected async submit(): Promise<void> {
+    const itemToSubmit = this.options[this.listIndex];
+
+    if (Select.itemIsBackButton(itemToSubmit)) {
+      const previousLevel = this.#parentCategories.pop();
+      if (typeof previousLevel === "object") {
+        this.options = previousLevel.options;
+        if (this.isCurrentlyInsideCategory()) {
+          this.listIndex = 2;
+        } else {
+          this.listIndex = 0;
+        }
+      }
+    } else if (Select.itemIsCategory(itemToSubmit)) {
+      this.#parentCategories.push({
+        options: this.options,
+        selectedCategoryIndex: this.listIndex,
+      });
+      this.options = this.itemsInsideCategory(itemToSubmit);
+      this.listIndex = 2;
+    } else {
+      await super.submit();
+    }
+  }
+
+  private itemsInsideCategory(
+    option: SelectOptionSettings,
+  ): SelectValueSettings {
+    return [
+      backItem,
+      ...Select.mapOptions([Select.separator(), ...option.options]),
+    ];
+  }
+
+  protected body(): string | Promise<string> {
+    let categoryText = "";
+
+    switch (this.#parentCategories.length) {
+      case 0:
+        return super.body();
+      case 1:
+        categoryText = dim(italic(this.nameOfParentCategory(0)));
+        break;
+
+      case 2: 
+        categoryText = dim(
+          italic(
+            this.nameOfParentCategory(0) + "/" + this.nameOfParentCategory(1),
+          ),
+        );
+        break;
+      
+      default: 
+        categoryText = dim(
+          italic(
+            this.nameOfParentCategory(0) + "/../" +
+              this.nameOfParentCategory(this.#parentCategories.length - 1),
+          ),
+        );
+        break;
+      
+    }
+
+    return categoryText + "\n" + super.body();
+  }
+
+  private nameOfParentCategory(index: number): string {
+    const { options, selectedCategoryIndex } = this.#parentCategories[index];
+    return options[selectedCategoryIndex].name;
+  }
+
+  protected match(): void {
+    const input: string = this.getCurrentInputValue().toLowerCase();
+
+    if (input === "") {
+      if (this.isCurrentlyInsideCategory()) {
+        const nearestParentCategories =
+          this.#parentCategories[this.#parentCategories.length - 1];
+        const categoryToRevertTo =
+          nearestParentCategories.options[nearestParentCategories.selectedCategoryIndex];
+
+        this.options = this.itemsInsideCategory(categoryToRevertTo);
+        this.listIndex = 2;
+      } else {
+        this.options = this.settings.options.slice();
+        this.clampListIndex();
+      }
+    } else {
+      const newOptions = this.itemsToSearchIn()
+        .filter((option) => Select.matchesItemOrSubItems(input, option));
```

Comment:
Of course, this code now means select options will never be sorted, which could be seen as a regression. So we could detect if there are no categories in the search results, and then sort only when that's the case.

We can also add sorting between options that are in the same category, but trying to sort all search results across different categories seems like it will just be confusing.

Reply:
We could try to sort all items by string distance and all category names alphabetically below the items.

Initial:
```
? Pick a color ⌕ 
group1
  matchedItem2
  matchedItem1
matchedItem2
matchedItem1
group2
  matchedItem1
  matchedItem2
```

After search:
```
? Pick a color ⌕ matchedItem
matchedItem1
matchedItem2
group1
  matchedItem1
  matchedItem2
group2
  matchedItem1
  matchedItem2
```

Reply:
but maybe we should put the sorting in general behind an option 🤔

Reply:
I think that risks putting the most relevant results the furthest away (if say `group2` contains an exact match). Maybe something like this is better:
- Divide the search results into "groups", where:
  - Any non-category item is part of the group of items that share the same parent category
  - Any non-category item that does not have a parent category becomes its own group, with no other members
- Define a group's distance to the search string as the smallest distance of all its group members to the input string, and sort the groups by this distance
- Internally in a group, also sort members by their distance to the input, relative to each other.

In other words, the example might look like this:
Initial:
```
? Pick a color ⌕ 
group1
  aaa
  abb
abb
aab
group2
  bbb
  aac
```

After search:
```
? Pick a color ⌕ aa
group1
  aaa
  abb
aab
group2
  aac
abb
```

Reply:
And without sorting, the ordering becomes essentially random, which I also don't know if is desirable. So I'm not sure if it's worth allowing to turn it off 🤷

Reply:
I think we can implement the sorting for now and look how it feels. If it feels confusing we can add an option to disable group and general sorting.

Reply:
Do you have a preference for which sorting to implement? 🙂

Reply:
I think you can just sort everithing for now, groups and entries.